### PR TITLE
feat(CalibrationClassification): support calibration classification to use t4dataset for train/val/test

### PR DESCRIPTION
**Pull request by vividf**
## Summary
This PR introduces **full support for T4Dataset** in the **Calibration Classification pipeline** of AWML. It enables streamlined **training, validation, and testing** using T4-format datasets, with a modular, camera-selectable data preparation and transform pipeline suitable for real-world deployment and large-scale experiments.

### 1. T4Dataset Support for Calibration Classification

- Added support for loading and using **T4-format datasets**.
- Introduced a new script:  
  `tools/calibration_classification/create_data_t4dataset.py`  
  to generate calibration classification info pickle files (`info.pkl`).
- Default: uses **all six cameras**; users can filter by specific camera channels via CLI.

#### Example Command
```bash
python3 tools/calibration_classification/create_data_t4dataset.py   --config /workspace/autoware_ml/configs/calibration_classification/dataset/t4dataset/gen2_base.py   --version gen2_base   --root_path ./data/t4dataset   -o ./data/t4dataset/calibration_info/
```

#### Example Output Format (`info.pkl`)
```python
{
  'frame_id': 'CAM_FRONT',
  'image': {
    'cam2ego': [...],
    'cam2img': [...],
    'img_path': 'data/camera/front/00003.jpg',
    'lidar2cam': [...],
    ...
  },
  'lidar_points': {
    'lidar_path': 'data/lidar/00003.pcd.bin',
    'lidar2ego': [...],
    ...
  },
  ...
}
```

---

#### 2.1. Pre-Training Calibration Visualization
```bash
python tools/calibration_classification/visualize_lidar_camera_projection.py --info_pkl data/t4dataset/calibration_info/t4dataset_x2_calib_infos_test.pkl --data_root data/t4dataset --output_dir ./calibration_visualization
```
- Added visualization to **verify calibration accuracy** before training.
- Projects LiDAR points onto camera images using `info.pkl` data.

#### 2.2. Runtime Visualization During Training / Validation / Testing

- Visualize both:
  - **Calibration (Projection)**
  - **Classification Results**
- Supports runtime debugging and interpretability.

#### 2.3. Visualization Types

| Type                   | Directory               | Description                                   |
|------------------------|--------------------------|-----------------------------------------------|
| **Projection Visualization** | `projection_vis_dir`     | Shows LiDAR points projected on camera image |
| **Results Visualization**    | `results_vis_dir`        | Shows model predictions on images            |

#### 2.4. Example Config
```python
# In your config file
projection_vis_dir = "./projection_vis_t4dataset/"
results_vis_dir = "./results_vis_t4dataset/"

# In transform pipeline
dict(
    type="CalibrationClassificationTransform",
    mode="validation",
    undistort=True,
    enable_augmentation=False,
    data_root=data_root,
    projection_vis_dir=projection_vis_dir,
    results_vis_dir=results_vis_dir,
),
```

---

### 3. Recommended Usage Strategy

| Phase              | Strategy                                                                 |
|--------------------|--------------------------------------------------------------------------|
| **Training/Validation**        | Disable visualization for speed (`projection_vis_dir=None`)             |
| **Test** | Enable both visualizations to analyze calibration and prediction quality |

#### Output Examples

- `projection_vis_dir/`: Projected LiDAR → image overlays  
- `results_vis_dir/`: Model predictions with class labels on images  

## Note

For usage examples and more details, please refer to the documentation in the following [link](https://github.com/vividf/AWML/tree/feat/calibration_classification_dataset/tools/calibration_classification)

## Test performed
Able to use t4dataset for calibration classification

Some examples
results_test_sample_185_00030_CAM_FRONT_RIGHT_pred_label_0_gt_label_0

results_test_sample_195_00032_CAM_FRONT_pred_label_1_gt_label_1

### CODE REVIEW

File: projects/CalibrationStatusClassification/deploy/main.py [hunk 1]
```diff
@@ -15,22 +15,22 @@
...
 import time
-from typing import Dict, Optional, Tuple
+from typing import Any, Dict, Optional, Tuple
 
 import mmengine
 import numpy as np
 import onnx
 import onnxruntime as ort
 import onnxsim
-import pycuda.autoinit
```
Comment (amadeuszsz) [inline projects/CalibrationStatusClassification/deploy/main.py:26]: Why this import has been removed?
https://documen.tician.de/pycuda/util.html

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
+_base_ = [
```
Comment (amadeuszsz) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:1]: Do we need to change suffix from `t4base` to `j6gen2`? I don't think we need to keep config separations for different sensor kits in this specific task.

File: tools/calibration_classification/README.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+## 2. Prepare dataset
+
+Prepare the dataset you use.
+
+### 2.1. T4dataset
+
+- Run docker
+
+```sh
+docker run -it --rm --gpus '"device=0"' --shm-size=64g --name awml -p 6006:6006 -v $PWD/:/workspace -v $PWD/data:/workspace/data autoware-ml
```
Comment (amadeuszsz) [inline tools/calibration_classification/README.md:26]: By default, I suggest to not specify device in documentation, just use all.

**Review by amadeuszsz — COMMENTED** [review]
Thanks for this work! Brief comments just from quick look on the PR. Later I will be able to check functionality and code itself.

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py (hunk 2 above)
Comment (vividf) [reply projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:1]: If we run this for other vehicle, they need to change to difference dataset

File: projects/CalibrationStatusClassification/deploy/main.py (hunk 1 above)
Comment (vividf) [reply projects/CalibrationStatusClassification/deploy/main.py]: Right! Thanks for the check! 
I accidentally removed it because my editor shows it is not used.
Addressed in https://github.com/tier4/AWML/pull/82/commits/08f317ce0e74758b4909bb2ac80f1758acf70d0a

**Comment by KSeangTan** [comment]
Hey, thanks for the work and just a quick comment 

I would suggest not to send a huge PR (> 500 lines, excluding configs/dataset id updates/documentations) if possible in the future, it makes reviewing process more difficult, and it takes more time for reviewers to find the focus.

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        type="PackInputs",
+        input_key="fused_img",
+        # meta_keys=["img_path", "fused_img", "image", "sample_idx", "frame_id", "frame_idx"],  # Visualization
+    ),
+]
+
+val_dataloader = dict(
+    batch_size=batch_size,
+    num_workers=num_workers,
+    persistent_workers=False,
```
Comment (amadeuszsz) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:101]: Is False here on purpose (opposite to train dataloader)?

Comment (vividf) [reply projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:101]: @amadeuszsz 
Yes,  it used `False` previously as well, since validation runs less frequently and doesn't benefit much from persistent workers (also to avoid unnecessary resource usage)

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    dataset=dict(
+        type="T4CalibrationClassificationDataset",
+        ann_file=info_directory_path + _base_.info_val_file_name,
+        pipeline=val_pipeline,
+        data_root=data_root,
+    ),
+)
+
+val_evaluator = dict(topk=(1,), type="mmpretrain.evaluation.Accuracy")
+
```
Comment (amadeuszsz) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:112]: Did we miss `test_cfg()` definition?

```suggestion

test_cfg = dict()

```

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    "mmpretrain::_base_/default_runtime.py",
+    "mmpretrain::_base_/models/resnet18.py",
+    "mmpretrain::_base_/schedules/imagenet_bs256.py",
+    "../../../../autoware_ml/configs/calibration_classification/dataset/t4dataset/gen2_base.py",
+]
+
+custom_imports = dict(
+    imports=["autoware_ml.calibration_classification.datasets.transforms.calibration_classification_transform"],
+    allow_failed_imports=False,
+)
```
Comment (amadeuszsz) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:11]: I added this accidentally, please remove it 🙏🏻 

```suggestion
```

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    val_hook_config = dict(type="ResultVisualizationHook", results_vis_dir=val_results_vis_dir, phases=["val"])
+    if data_root is not None:
+        val_hook_config["data_root"] = data_root
+    custom_hooks.append(val_hook_config)
+
+if test_results_vis_dir is not None:
+    test_hook_config = dict(type="ResultVisualizationHook", results_vis_dir=test_results_vis_dir, phases=["test"])
+    if data_root is not None:
+        test_hook_config["data_root"] = data_root
+    custom_hooks.append(test_hook_config)
```
[the remark this answers is not in the document]
Comment (vividf) [reply projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py]: Great idea, thanks.
Addressed in https://github.com/tier4/AWML/pull/82/commits/e66d5de4a89b7a14310a2974d06c12ed9c8bbc89

File: projects/CalibrationStatusClassification/deploy/main.py [hunk 8]
```diff
@@ -41,6 +42,104 @@
...
+
+    # Validate sample structure
+    required_keys = ["image", "lidar_points"]
+    if not all(key in sample for key in required_keys):
+        raise ValueError(f"Sample {sample_idx} has invalid structure. Required keys: {required_keys}")
+
+    return sample
+
+
+def load_sample_data_from_info_pkl(
```
Comment (KSeangTan) [inline projects/CalibrationStatusClassification/deploy/main.py:94]: Do we need to load an image for deployment? We can just make a random image if we know the shape right?

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 9]
```diff
@@ -0,0 +1,864 @@
...
+    def __post_init__(self):
+        """Initialize new_camera_matrix if not provided.
+        Sets new_camera_matrix to a copy of camera_matrix if it's None.
+        """
+        if self.new_camera_matrix is None:
+            self.new_camera_matrix = self.camera_matrix.copy()
+
+
+@TRANSFORMS.register_module()
+class CalibrationClassificationTransform(BaseTransform):
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:53]: Understand, however, for this PR, I will keep it this way since Amadeusz will also make some updates afterward.

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 10]
```diff
@@ -0,0 +1,864 @@
...
+            image: Input camera image.
+            lidar_data: LiDAR point cloud and intensity data.
+            calibration_data: Camera calibration parameters.
+            augmentation_tf: Optional augmentation transformation matrix.
+            use_lidar2cam: Whether to use direct lidar2cam transformation.
+        Returns:
+            Combined image with RGB, depth, and intensity channels.
+        """
+        points = lidar_data["pointcloud"]
+        N = points.shape[0]
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Thanks, addressed in https://github.com/tier4/AWML/pull/82/commits/b44cd9c5bd3481dd93c9dadd214b3ebcb5f7c7d9

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 11]
```diff
@@ -0,0 +1,864 @@
...
+    ) -> Tuple[np.ndarray, CalibrationData]:
+        """Undistort image and update calibration data.
+        Args:
+            calibration_data: Camera calibration parameters.
+            image: Input image to undistort.
+            alpha: Free scaling parameter for undistortion.
+        Returns:
+            Tuple of undistorted image and updated calibration data.
+        """
+        if np.any(calibration_data.distortion_coefficients):
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Thanks, addressed in https://github.com/tier4/AWML/pull/82/commits/208fc3c580d0d6e19f6b9b9bbb7cc552928f7f52

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 12]
```diff
@@ -0,0 +1,864 @@
...
+
+
+@dataclass
+class CalibrationData:
+    """Structured representation of camera calibration data.
+    This class holds all the necessary calibration information for camera-LiDAR
+    coordinate transformations and image processing operations.
+    """
+
+    camera_matrix: np.ndarray  # Original camera intrinsic matrix (3x3)
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Fixed in https://github.com/tier4/AWML/pull/82/commits/b54467131d56902cf256c817d47932cf51306eb5

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 13]
```diff
@@ -0,0 +1,864 @@
...
+    def transform(self, results: Dict[str, Any], force_generate_miscalibration: bool = False) -> Dict[str, Any]:
+        """Transform input data for calibration classification.
+        Args:
+            results (Dict[str, Any]): Input data dictionary containing all info.pkl fields.
+            force_generate_miscalibration (bool): Whether to force generation of miscalibration. Defaults to False.
+        Returns:
+            Dict[str, Any]: Transformed data dictionary with processed images and labels.
+        """
+        logger.debug(f"Starting transform for sample {results.get('sample_idx', 'unknown')}")
+        self._set_random_seeds(results)
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Thanks! move them to config https://github.com/tier4/AWML/pull/82/commits/0501636d70321fd009d0527c74e7f99868759c1d

File: projects/CalibrationStatusClassification/deploy/main.py [hunk 14]
```diff
@@ -41,6 +42,104 @@
...
+
+    # Create transform for deployment
+    mode = "test" if not force_generate_miscalibration else "train"
+    transform = CalibrationClassificationTransform(
+        mode=mode,
+        data_root=data_root,
+        projection_vis_dir=None,
+        results_vis_dir=None,
+        enable_augmentation=False,
+    )
```
Comment (amadeuszsz) [inline projects/CalibrationStatusClassification/deploy/main.py:130]: This definition is missing `transform_config`.
According to `transform_config` in `gen2_base.py` - I think this dictionary does not fit to the default configuration because base configuration file consists only "irrelevant" parameters for training loop. Pre-processing affects training loop significantly, therefore I see it as a simple class arguments. In other words each `transform_config=_base_.transform_config` occurrence just simply replace with `crop_ratio=0.6` etc...

We want to explicitly define it in configuration file. I checked other projects and they follow same approach.

File: tools/calibration_classification/create_data_t4dataset.py [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+        # Skip invalid sample data (is_valid = false)
+        if sd.get("is_valid") is False:
+            invalid_count += 1
+            continue
+        frame_idx: str = extract_frame_index(sd["filename"])
+        frame_groups[frame_idx].append(sd)
+
+    logger.warning(
+        f"Found {len(frame_groups)} frames in scene {scene_id} (filtered out {invalid_count} invalid samples)"
+    )
```
[the remark this answers is not in the document]
Comment (vividf) [reply tools/calibration_classification/create_data_t4dataset.py]: Thanks for the check.
Addressed in https://github.com/tier4/AWML/pull/82/commits/a8e8508843ca883004c5b0be6eff5135b41850c3

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 16]
```diff
@@ -0,0 +1,816 @@
+import os
+import random
+from dataclasses import dataclass
+from enum import Enum
+from tkinter import NO
```
[the remark this answers is not in the document]
Comment (vividf) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Thanks!
Addressed in https://github.com/tier4/AWML/pull/82/commits/4b51fbb089e4494d14261c83e676534b44c01b88

File: projects/CalibrationStatusClassification/deploy/main.py (hunk 14 above)
Comment (vividf) [reply projects/CalibrationStatusClassification/deploy/main.py]: Thanks for the comment, fix them in https://github.com/tier4/AWML/pull/82/commits/74a56f0d94ea7d63f32bcb3945a4cf3d70a643f4.

I agree that pre-processing can significantly affect the training loop, and making such parameters explicit helps with transparency during experiments.
However, I believe keeping the transform configuration in gen2_base.py is still beneficial for consistency and maintainability, especially since for the same dataset, the transform settings are usually identical — for example, both ResNet18 and ResNet50 would use the same configuration.
As a compromise, we could keep the common defaults in gen2_base.py for shared use, while allowing overrides through class arguments when needed. This way, constants like crop_ratio or max_distortion — which were previously hardcoded and rarely changed — remain centralized, but can still be explicitly set in the class call for experiments that require variations.

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 17]
```diff
@@ -0,0 +1,815 @@
...
+
+        # Generate input data
+        input_data = self._generate_input_data(augmented_image, lidar_data, augmented_calibration, augmentation_tf)
+        logger.debug(f"Generated input data shape: {input_data.shape}")
+
+        # Visualization
+        if self.projection_vis_dir is not None:
+            self._visualize_projection(input_data, label, results, self.mode.value)
+
+        results["fused_img"] = input_data
```
Comment (amadeuszsz) [inline autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:204]: All the functions in this scope consist of `float32` type annotation for image array. However, image keeps `uint8t` type. Could you please confirm it?

**Review by amadeuszsz — APPROVED** [review]
Thank you for this PR! Dataset issue has been fully addressed 🎉 

We would like to add more changes to this project, but they are out of dataset scope. Therefore, separate PR will be issued soon.

Referring to dataset, we might want to consider velocity filter as some frames characterize with blur and image & cloud misalignment. We can discuss it later.
